### PR TITLE
[Merged by Bors] - implement `Reflect` for `Input<T>`, some misc improvements to reflect value derive

### PR DESCRIPTION
**Pull request by jakobhellermann**
# Objective

- I'm currently working on being able to call methods on reflect types (https://github.com/jakobhellermann/bevy_reflect_fns)
- for that, I'd like to add methods to the `Input<KeyCode>` resource (which I'm doing by registering type data)
- implementing `Reflect` is currently a requirement for having type data in the `TypeRegistry`

## Solution

- derive `Reflect` for `KeyCode` and `Input`
- uses `#[reflect_value]` for `Input`, since it's fields aren't supposed to be observable
- using reflect_value would need `Clone` bounds on `T`, but since all the methods (`.pressed` etc) already require `T: Copy`, I unified everything to requiring `Copy`
- add `Send + Sync + 'static` bounds, also required by reflect derive

## Unrelated improvements 
I can extract into a separate PR if needed.

- the `Reflect` derive would previously ignore `#[reflect_value]` and only accept `#[reflect_value()]` which was a bit confusing
- the generated code used `val.clone()` on a reference, which is fine if `val` impls `Clone`, but otherwise also compiles with a worse error message. Change to `std::clone::Clone::clone(val)` instead which gives a neat `T does not implement Clone` error

### CODE REVIEW

**Review by NiklasEi — APPROVED** [review]

**Review by devil-ira — COMMENTED** [review]
It would be nice if  `ScanCode` was also `Reflect` : )

**Review by devil-ira — APPROVED** [review]
Not familiar with the derive but LGTM.

File: crates/bevy_input/src/keyboard.rs [hunk 1]
```diff
@@ -60,7 +61,7 @@ pub fn keyboard_input_system(
 /// ## Updating
 ///
 /// The resource is updated inside of the [`keyboard_input_system`](crate::keyboard::keyboard_input_system).
-#[derive(Debug, Hash, Ord, PartialOrd, PartialEq, Eq, Clone, Copy)]
+#[derive(Reflect, FromReflect, Debug, Hash, Ord, PartialOrd, PartialEq, Eq, Clone, Copy)]
```
Comment (MrGVSV) [inline crates/bevy_input/src/keyboard.rs:64]: Any reason to not register reflected traits like `PartialEq` and `Hash`?

(Also on `ScanCode`)

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs [hunk 2]
```diff
@@ -111,19 +111,18 @@ impl<'a> ReflectDerive<'a> {
...
                     traits = ReflectTraits::from_nested_metas(&meta_list.nested);
-                } else if ident == REFLECT_VALUE_ATTRIBUTE_NAME {
+                }
+                Meta::List(meta_list) if meta_list.path.is_ident(REFLECT_VALUE_ATTRIBUTE_NAME) => {
                     force_reflect_value = true;
                     traits = ReflectTraits::from_nested_metas(&meta_list.nested);
                 }
+                Meta::Path(path) if path.is_ident(REFLECT_VALUE_ATTRIBUTE_NAME) => {
+                    force_reflect_value = true;
+                }
```
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:124]: Nice! Didn't even realize this was broken like that 😅

**Review by MrGVSV — APPROVED** [review]

File: crates/bevy_input/src/keyboard.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_input/src/keyboard.rs:64]: @jakobhellermann, I'd like to address this (either way) before I merge. This is the only remaining blocker IMO.

**Pull request closed without merging**